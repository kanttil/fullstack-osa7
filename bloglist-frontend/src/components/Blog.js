import React from 'react'
import PropTypes from 'prop-types'
import { Button, Form, Table } from 'semantic-ui-react'

const DeleteButton = ({ blog, user, deleteBlog }) => {
    if (blog.user === undefined || blog.user.username === user.username) {
        return (
            <Button onClick={deleteBlog(blog)}>delete</Button>
        )
    }

    return null
}

class Blog extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            comment: ''
        }
    }

    deleteBlog = (blog) => () => {
        if (window.confirm(`delete '${blog.title}' by ${blog.author}?`)) {
            this.props.removeBlog(blog._id)
        }
    }

    likeBlog = (blog) => () => {
        const blogObject = {
            title: blog.title,
            author: blog.author,
            url: blog.url,
            likes: blog.likes + 1,
            user: blog.user._id
        }

        this.props.updateBlog(blog._id, blogObject)
    }

    handleCommentFieldChange = (event) => {
        event.preventDefault()
        this.setState({ comment: event.target.value })
    }

    submitComment = (event) => {
        event.preventDefault()

        const commentObject = {
            content: this.state.comment,
            blog: this.props.blog._id
        }
        this.props.addComment(this.props.blog, commentObject)

        this.setState({ comment: '' })
    }

    render() {
        const blog = this.props.blog
        if (blog === undefined) return null

        let anonymous = null
        if (blog.user === undefined) {
            anonymous = 'anonymous'
        }

        const titleStyle = {
            color: 'dimGray'
        }

        return (
            <div>
                <Table>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell>
                                <h2 style={titleStyle}>{blog.title} {blog.author}</h2>
                            </Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        <Table.Row>
                            <Table.Cell>
                                <a href={blog.url}>{blog.url}</a>
                            </Table.Cell>
                        </Table.Row>
                        <Table.Row>
                            <Table.Cell>
                                {blog.likes} likes &nbsp;
                                <Button onClick={this.likeBlog(blog)}>like</Button>
                            </Table.Cell>
                        </Table.Row>
                        <Table.Row>
                            <Table.Cell>
                                <div>added by {anonymous || blog.user.name}</div>
                            </Table.Cell>
                        </Table.Row>
                    </Table.Body>
                </Table>
                <DeleteButton blog={blog} user={this.props.user} deleteBlog={this.deleteBlog} />

                <h3>comments</h3>
                <Form onSubmit={this.submitComment}>
                    <Form.Field inline>
                        <input
                            type="text"
                            name="comment"
                            value={this.state.comment}
                            onChange={this.handleCommentFieldChange} />
                        <Button type="submit">add comment</Button>
                    </Form.Field>
                </Form>
                <ul>
                    {blog.comments.map(comment =>
                        <li key={comment._id}>{comment.content}</li>)}
                </ul>
            </div>
        )
    }
}

Blog.propTypes = {
    removeBlog: PropTypes.func.isRequired,
    updateBlog: PropTypes.func.isRequired,
    addComment: PropTypes.func.isRequired
}

export default Blog