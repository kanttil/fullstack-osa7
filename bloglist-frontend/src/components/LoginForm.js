import React from 'react'
import PropTypes from 'prop-types'
import { Form, Button } from 'semantic-ui-react'

class LoginForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            username: '',
            password: ''
        }
    }

    handleLoginFieldChange = (event) => {
        this.setState({ [event.target.name]: event.target.value })
    }

    onSubmit = (event) => {
        event.preventDefault()
        this.props.login(this.state.username, this.state.password)
        this.setState({
            username: '',
            password: ''
        })
    }

    render() {
        const titleStyle = {
            color: 'dimGray'
        }

        return (
            <div>
                <h2 style={titleStyle}>Log in to application</h2>

                <Form onSubmit={this.onSubmit}>
                    <Form.Field>
                        <label>username:</label>
                        <input
                            type="text"
                            name="username"
                            value={this.state.username}
                            onChange={this.handleLoginFieldChange}
                        />
                    </Form.Field>
                    <Form.Field>
                        <label>password:</label>
                        <input
                            type="password"
                            name="password"
                            value={this.state.password}
                            onChange={this.handleLoginFieldChange}
                        />
                    </Form.Field>
                    <Button type="submit">login</Button>
                </Form>
            </div>
        )
    }
}

LoginForm.propTypes = {
    login: PropTypes.func.isRequired
}

export default LoginForm