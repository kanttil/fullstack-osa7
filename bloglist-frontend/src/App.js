import React from 'react'
import BlogList from './components/BlogList'
import Blog from './components/Blog'
import Togglable from './components/Togglable'
import BlogForm from './components/BlogForm'
import LoginForm from './components/LoginForm'
import UserList from './components/UserList'
import User from './components/User'
import blogService from './services/blogs'
import userService from './services/users'
import loginService from './services/login'

import { BrowserRouter as Router, Route, Link } from 'react-router-dom'
import { Container, Button, Menu } from 'semantic-ui-react'

const Notification = ({ message, color }) => {
    if (message === null) {
        return null
    }

    const notificationStyle = {
        color,
        border: 'solid',
        borderWidth: 2,
        borderRadius: 10,
        padding: 10,
        marginBottom: 10
    }

    return (
        <div style={notificationStyle}>
            {message}
        </div>
    )
}

class App extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            blogs: [],
            users: [],
            notification: null,
            notificationColor: null,
            user: null
        }
    }

    componentDidMount() {
        blogService.getAll().then(blogs =>
            this.setState({ blogs })
        )
        userService.getAll().then(users =>
            this.setState({ users })
        )
        const loggedUserJSON = window.localStorage.getItem('loggedUser')
        if (loggedUserJSON) {
            const user = JSON.parse(loggedUserJSON)
            blogService.setToken(user.token)
            this.setState({ user })
        }
    }

    login = async (username, password) => {
        try {
            const user = await loginService.login({
                username,
                password
            })

            window.localStorage.setItem('loggedUser', JSON.stringify(user))
            blogService.setToken(user.token)
            this.setState({
                user,
                notification: `${user.name} logged in`,
                notificationColor: 'green'
            })
            setTimeout(() => {
                this.setState({ notification: null })
            }, 3000)
        } catch (exception) {
            this.setState({
                notification: 'wrong username or password',
                notificationColor: 'red'
            })
            setTimeout(() => {
                this.setState({ notification: null })
            }, 3000)
        }
    }

    logout = (event) => {
        event.preventDefault()

        window.localStorage.clear()
        blogService.setToken(null)
        this.setState({
            user: null,
            notification: 'logged out',
            notificationColor: 'green'
        })
        setTimeout(() => {
            this.setState({ notification: null })
        }, 3000)
    }

    addBlog = async (blogObject) => {
        this.blogForm.toggleVisibility()

        const newBlog = await blogService.create(blogObject)
        const blogs = await blogService.getAll()
        const users = await userService.getAll()
        this.setState({
            blogs,
            users,
            notification: `a new blog '${newBlog.title}' by ${newBlog.author} added`,
            notificationColor: 'green'
        })
        setTimeout(() => {
            this.setState({ notification: null })
        }, 3000)
    }

    updateBlog = async (id, blogObject) => {
        await blogService.update(id, blogObject)
        const blogs = await blogService.getAll()

        this.setState({ blogs })
    }

    removeBlog = async (id) => {
        await blogService.remove(id)
        this.setState({ blogs: this.state.blogs.filter(blog => blog._id !== id) })
    }

    addComment = async (blog, commentObject) => {
        const newComment = await blogService.createComment(blog._id, commentObject)
        const blogs = await blogService.getAll()
        this.setState({
            blogs,
            notification: `comment '${newComment.content}' added to blog '${blog.title}'`,
            notificationColor: 'green'
        })
        setTimeout(() => {
            this.setState({ notification: null })
        }, 3000)
    }

    render() {
        const blogById = (id) => this.state.blogs.find(blog => blog._id === id)
        const userById = (id) => this.state.users.find(user => user._id === id)

        const titleStyle = {
            color: 'cornFlowerBlue'
        }

        if (this.state.user === null) {
            return (
                <Container>
                    <Notification
                        message={this.state.notification}
                        color={this.state.notificationColor} />
                    <LoginForm login={this.login} />
                </Container>
            )
        }

        return (
            <Container>
                <Router>
                    <div>
                        <Notification
                            message={this.state.notification}
                            color={this.state.notificationColor} />
                        <h1 style={titleStyle}>blog app</h1>
                        <Menu>
                            <Menu.Item link>
                                <Link to="/">blogs</Link>
                            </Menu.Item>
                            <Menu.Item link>
                                <Link to="/users">users</Link>
                            </Menu.Item>
                            <Menu.Item>
                                {this.state.user.name} logged in &nbsp;
                                <Button onClick={this.logout}>logout</Button>
                            </Menu.Item>
                        </Menu>
                        <Togglable buttonLabel="create new blog" ref={component => this.blogForm = component}>
                            <BlogForm addBlog={this.addBlog} />
                        </Togglable>
                        <Route exact path="/" render={() =>
                            <BlogList blogs={this.state.blogs} />} />
                        <Route exact path="/users" render={() =>
                            <UserList users={this.state.users} />} />
                        <Route exact path="/users/:id" render={({ match }) =>
                            <User user={userById(match.params.id)} />} />
                        <Route exact path="/blogs/:id" render={({ match }) =>
                            <Blog
                                blog={blogById(match.params.id)}
                                user={this.state.user}
                                updateBlog={this.updateBlog}
                                removeBlog={this.removeBlog}
                                addComment={this.addComment} />} />
                    </div>
                </Router>
            </Container>
        )
    }
}

export default App