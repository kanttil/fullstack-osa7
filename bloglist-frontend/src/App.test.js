import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import App from './App'
jest.mock('./services/blogs')
jest.mock('./services/users')
import blogService from './services/blogs'
import LoginForm from './components/LoginForm'
import Blog from './components/Blog'

describe('<App />', () => {
    let app

    describe('when user is not logged', () => {
        beforeEach(() => {
            app = mount(<App />)
        })

        it('only login form is rendered', () => {
            app.update()
            const blogComponents = app.find('.blogLink')
            const loginForm = app.find(LoginForm)

            expect(blogComponents.length).toBe(0)
            expect(loginForm.text()).toBe(app.text())
        })

        it('it renders correctly', () => {
            const tree = renderer
                .create(<LoginForm login={() => null} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('when user is logged', () => {
        beforeEach(() => {
            const user = {
                username: 'tester',
                token: '1231231214',
                name: 'Testaaja'
            }

            localStorage.setItem('loggedUser', JSON.stringify(user))
            app = mount(<App />)
        })

        it('all blogs are rendered as links', () => {
            app.update()
            const blogLinks = app.find('.blogLink')

            expect(blogLinks.length).toBe(blogService.blogs.length)
        })

        it('at first only the title and author are visible', () => {
            app.update()
            const blogLink = app.find('.blogLink').first()
            const details = app.find('.details')
            const firstBlog = blogService.blogs[0]

            expect(blogLink.text()).toEqual(firstBlog.title + ' ' + firstBlog.author)
            expect(details.length).toBe(0)
        })

        it('clicking a link opens the details of a blog', () => {
            app.update()
            const blogLink = app.find('.blogLink').first()
            blogLink.simulate('click', { button: 0 })

            const details = app.find(Blog)
            const firstBlog = blogService.blogs[0]

            expect(details.length).toBe(1)
            expect(details.text()).toContain(firstBlog.url && firstBlog.likes)
        })

        it('the details render correctly', () => {
            const f = () => null
            const tree = renderer
                .create(<Blog
                    blog={blogService.blogs[0]}
                    user={{ username: 'testUser' }}
                    removeBlog={f}
                    updateBlog={f}
                    addComment={f} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })
})