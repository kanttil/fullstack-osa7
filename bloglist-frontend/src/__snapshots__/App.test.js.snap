// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when user is logged the details render correctly 1`] = `
<div>
  <table
    className="ui table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
        >
          <h2
            style={
              Object {
                "color": "dimGray",
              }
            }
          >
            React patterns
             
            Michael Chan
          </h2>
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className=""
        >
          <a
            href="https://reactpatterns.com/"
          >
            https://reactpatterns.com/
          </a>
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          1
           likes Â 
          <button
            className="ui button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            like
          </button>
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          <div>
            added by 
            User1
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <h3>
    comments
  </h3>
  <form
    action={undefined}
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="inline field"
    >
      <input
        name="comment"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="ui button"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
        type="submit"
      >
        add comment
      </button>
    </div>
  </form>
  <ul />
</div>
`;

exports[`<App /> when user is not logged it renders correctly 1`] = `
<div>
  <h2
    style={
      Object {
        "color": "dimGray",
      }
    }
  >
    Log in to application
  </h2>
  <form
    action={undefined}
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label>
        username:
      </label>
      <input
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        password:
      </label>
      <input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="ui button"
      disabled={undefined}
      onClick={[Function]}
      role="button"
      tabIndex={undefined}
      type="submit"
    >
      login
    </button>
  </form>
</div>
`;
